### ISSUE TITLE
run test_restore.py and got _pickle.UnpicklingError: state is not a dictionary

### ISSUE THREAD

**Issue by LinWencong** [microsoft/FLAML#542]
I use BlendSearch save() and restore() functions to achieve a warm start from a file, but got these problem。
so I run test_restore.py, and got the same problem 

  File "test_restore.py", line 57, in testWarmStart
    results_exp_2 = self.run_explicit_restore(r_state, checkpoint_path)
  File "test_restore.py", line 38, in run_explicit_restore
    search_alg2.restore(checkpoint_path)
  File "/usr/local/python3.6/lib/python3.6/site-packages/flaml/searcher/suggestion.py", line 252, in restore
    self.searcher.restore(checkpoint_path)
  File "/usr/local/python3.6/lib/python3.6/site-packages/flaml/searcher/blendsearch.py", line 330, in restore
    state = pickle.load(inputFile)
_pickle.UnpicklingError: state is not a dictionary

Would you mind telling me it's the way I'm using it wrong? Or is my understanding of what this function does wrong?
Thank you so much for your attention to this issue❤️

env: ubuntu 16.04  & FLAML==0.10.0

**Comment by qingyun-wu** [comment]
Hi @[LinWencong](https://github.com/LinWencong),

Thank you for raising the issue and making a contribution to resolving it. I have two questions/comments:
(1) I cannot reproduce the error you reported by running test_restore.py. Can you provide more information about how you were running the test? And can you also try the latest flaml version v1.0.2 (although I did not get the error in v0.10.0 as well).
I am running the test via 
```
pytest test/tune/test_restore.py
```
using the latest code here: https://github.com/qingyun-wu/FLAML/blob/test-restore/test/tune/test_restore.py
where I added an additional test for BlendSearch (in the original file, it only has a test for CFO).

(2) The changes in your PR make `flaml.tune` and `ray.tune` not consistent which failed one of our tests: https://github.com/microsoft/FLAML/blob/main/test/tune/test_flaml_raytune_consistency.py
 
If your changes are necessary after we have a conclusion about (1), you will need to revise the PR to make `flaml.tune` and `ray.tune`. I will try to help there if the changes are necessary.

Thank you!

**Comment by LinWencong** [comment]
Referring to the error message of Build / build (ubuntu-latest, 3.7), it may be wrong to simply remove the random state without saving the complete information;
But I don't understand why I can pass several of these test pipelines

**Comment by LinWencong** [comment]
Referring to the error message of Build / build (ubuntu-latest, 3.7), it may be wrong to simply remove the random state without saving the complete information;
But I don't understand why I can pass several of these test pipelines

**Comment by LinWencong** [comment]
Hi  @qingyun-wu,

Thanks for your guidance❤️.

Referring to the error message of test pipeline : Build / build (ubuntu-latest, 3.7), I found the inconsistency problem as well.
It may be incorrect to simply remove rs_random without saving its entire state when saving. When restoring, rs_random is initialized to a new state, and no historical information is recorded, which will also cause inconsistencies.

But I'm confused about the part of testPipeline that I pass.

I found a way to reproduce the error.

this is my env:

ubuntu18.04
python3.6.7
flaml==0.10.0
ray==1.12.0
optuna==2.8.0
numpy==1.16.4

I am running the test via

`pytest test_restore.py`

and got 

```
...

test_restore.py:59:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test_restore.py:40: in run_explicit_restore
    search_alg2.restore(checkpoint_path)
../lib/python3.6/site-packages/flaml/searcher/suggestion.py:252: in restore
    self.searcher.restore(checkpoint_path)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <flaml.searcher.blendsearch.CFO object at 0x7fe367cec668>, checkpoint_path = '/tmp/tmpq07n16n4/searcher-state-Test.pkl'

    def restore(self, checkpoint_path: str):
        """restore states from checkpoint."""
        with open(checkpoint_path, "rb") as inputFile:
>           state = pickle.load(inputFile)
E           _pickle.UnpicklingError: state is not a dictionary

...

../lib/python3.6/site-packages/flaml/searcher/blendsearch.py:330: UnpicklingError

../lib/python3.6/site-packages/flaml/searcher/blendsearch.py:330: UnpicklingError
...
================================================================== short test summary info ===================================================================
FAILED test_restore.py::CFOWarmStartTest::testWarmStart - _pickle.UnpicklingError: state is not a dictionary
FAILED test_restore.py::BlendsearchWarmStartTest::testWarmStart - _pickle.UnpicklingError: state is not a dictionary
=========================================================== 2 failed, 2 passed, 1 warning in 2.21s ===========================================================

```
I can reproduce this error also in `numpy==1.16.*`

Besides, I found a way to fix it : `numpy>=1.17.0rc1`
I have verified it in several versions of numpy1.7.0 to 1.19.5, and passed the test.

In summary, is it possible to upgrade the version of numpy that satisfies flaml  (setup.py line 17 : `"NumPy>=1.16.2"`)?

It may be correct : `"NumPy>=1.7.0"`.

**Comment by LinWencong** [comment]
Maybe this part is causing the error.

/flaml/tune/sample.py 

```
# Backwards compatibility
try:
    # Added in numpy>=1.17 but we require numpy>=1.16
    np_random_generator = np.random.Generator
    LEGACY_RNG = False
except AttributeError:

    class np_random_generator:
        pass

    LEGACY_RNG = True
```

I will do more tests.

**Comment by sonichi** [comment]
Thanks @LinWencong for looking into this. If you can confirm that numpy version is the cause, I agree with updating the numpy version in setup.py.

**Comment by qingyun-wu** [comment]
@[LinWencong](https://github.com/LinWencong),

Thank you for looking into it. I tried "NumPy==1.16.2", and "NumPy==1.16.4" and indeed reproduced the error. With versions "NumPy>=1.17.0", the tests work well. I guess we can update the numpy version in setup.py.  @sonichi

**Comment by sonichi** [comment]
@LinWencong would you like to either update #545 or create a new PR to solve this issue?

**Comment by LinWencong** [comment]
> @LinWencong would you like to either update #545 or create a new PR to solve this issue?

I am glad to do that 😊. 
At the same time, I also want to explore the real cause of this problem with numpy<1.16.* . 
I will create a new PR soon.
